api/renter.go

```diff
@@ -350,49 +378,118 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
 
 // renterDownloadHandler handles the API call to download a file.
 func (api *API) renterDownloadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	params, err := api.parseDownloadParameters(w, req, ps)
+	if err != nil {
+		WriteError(w, Error{err.Error()}, http.StatusBadRequest)
 		return
 	}
 
-	err := api.renter.Download(strings.TrimPrefix(ps.ByName("siapath"), "/"), destination)
-	if err != nil {
-		WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
-		return
+	if params.Async { // Create goroutine if `async` param set.
+		// check for errors for 5 seconds to catch validation errors (no file with
+		// that path, invalid parameters, insufficient hosts, etc)
+		errchan := make(chan error)
+		go func() {
+			errchan <- api.renter.Download(params)
+		}()
+		select {
+		case err = <-errchan:
+			if err != nil {
+				WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
+				return
+			}
+		case <-time.After(time.Second * 5):
+		}
+	} else {
+		err := api.renter.Download(params)
+		if err != nil {
+			WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
+			return
+		}
 	}
 
-	WriteSuccess(w)
+	if params.Httpwriter == nil {
+		// `httpresp=true` causes writes to w before this line is run, automatically
+		// adding `200 Status OK` code to response. Calling this results in a
+		// multiple calls to WriteHeaders() errors.
+		WriteSuccess(w)
+		return
+	}
 }
 
 // renterDownloadAsyncHandler handles the API call to download a file asynchronously.
 func (api *API) renterDownloadAsyncHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	// Ensure that the async flag is not set to any other value than "true" or "".
+	async := req.FormValue("async")
+	if async == "false" || async != "true" && async != "" {
+		WriteError(w, Error{"/downloadasync require the `async` parameter to be either equal to `true` or be empty."}, http.StatusBadRequest)
```

Comment:
if you are calling the async endpoint and also setting the async param to something other  than "true", you've made a mistake. The API should catch things like that. Seems reasonable enough to me that it would happen.